File: j2kt/jre/java/common/java/lang/Float.java
/*
 *  Licensed to the Apache Software Foundation (ASF) under one or more
 *  contributor license agreements.  See the NOTICE file distributed with
 *  this work for additional information regarding copyright ownership.
 *  The ASF licenses this file to You under the Apache License, Version 2.0
 *  (the "License"); you may not use this file except in compliance with
 *  the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

package java.lang;

import static javaemul.internal.KtNativeUtils.ktNative;

import javaemul.internal.annotations.KtName;
import javaemul.internal.annotations.KtNative;
import org.jspecify.nullness.NullMarked;
import org.jspecify.nullness.Nullable;

@KtNative(name = "kotlin.Float")
@NullMarked
public final class Float extends Number implements Comparable<Float> {

  public static final float MAX_VALUE = 0x1.fffffeP+127f;

  public static final float MIN_VALUE = 0x0.000002P-126f;

  public static final float NaN = 0.0f / 0.0f;

  public static final float POSITIVE_INFINITY = 1.0f / 0.0f;

  public static final float NEGATIVE_INFINITY = -1.0f / 0.0f;

  public static final float MIN_NORMAL = 0x1.0p-126f;

  public static final int MAX_EXPONENT = 127;

  public static final int MIN_EXPONENT = -126;

  public static final Class<Float> TYPE = ktNative();

  @KtName("SIZE_BITS")
  public static final int SIZE = 32;

  @KtName("SIZE_BYTES")
  public static final int BYTES = SIZE / Byte.SIZE;

  public Float(float value) {}

  public Float(double value) {}

  public Float(String string) throws NumberFormatException {}

  @Override
  public native int compareTo(Float object);

  @KtName("toInt_toByte")
  @Override
  public native byte byteValue();

  @Override
  public native double doubleValue();

  @Override
  public native boolean equals(@Nullable Object object);

  public static native int floatToIntBits(float value);

  public static native int floatToRawIntBits(float value);

  @Override
  public native float floatValue();

  @Override
  public native int hashCode();

  @KtName("fromBits")
  public static native float intBitsToFloat(int bits);

  @Override
  public native int intValue();

  public native boolean isInfinite();

  public static native boolean isInfinite(float f);

  public static native boolean isFinite(float f);

  public native boolean isNaN();

  public static native boolean isNaN(float f);

  @Override
  public native long longValue();

  public static native float max(float a, float b);

  public static native float min(float a, float b);

  public static native float sum(float a, float b);

  public static native float parseFloat(String string) throws NumberFormatException;

  @KtName("toInt_toShort")
  @Override
  public native short shortValue();

  @Override
  public native String toString();

  public static native String toString(float f);

  public static native Float valueOf(String string) throws NumberFormatException;

  public static native int compare(float float1, float float2);

  public static native Float valueOf(float f);

  public static native String toHexString(float f);

  public static native int hashCode(float f);
}

File: j2kt/jre/java/native/java/lang/Float.kt
/*
 * Copyright 2022 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package java.lang

import kotlin.jvm.javaPrimitiveType

/** java.lang.Float static method emulations */
class Float {
  companion object {
    operator fun invoke(f: kotlin.Float): kotlin.Float = f

    operator fun invoke(s: kotlin.String): kotlin.Float = s.toFloat()

    const val MAX_VALUE = kotlin.Float.MAX_VALUE

    const val MIN_VALUE = kotlin.Float.MIN_VALUE

    const val NaN = kotlin.Float.NaN

    const val POSITIVE_INFINITY = kotlin.Float.POSITIVE_INFINITY

    const val NEGATIVE_INFINITY = kotlin.Float.NEGATIVE_INFINITY

    const val MIN_NORMAL: kotlin.Float = 1.1754943508222875e-38F // 0x1.0p-126f

    const val MAX_EXPONENT = 127

    const val MIN_EXPONENT = -126

    const val SIZE = kotlin.Float.SIZE_BITS

    const val BYTES = kotlin.Float.SIZE_BYTES

    val TYPE: Class<kotlin.Float> = kotlin.Float::class.javaPrimitiveType!!

    fun valueOf(f: kotlin.Float): kotlin.Float = f

    fun valueOf(s: kotlin.String): kotlin.Float = s.toFloat()

    fun compare(f1: kotlin.Float, f2: kotlin.Float): Int = f1.compareTo(f2)

    fun toString(f: kotlin.Float): kotlin.String = f.toString()

    fun parseFloat(s: kotlin.String): kotlin.Float = s.toFloat()

    fun hashCode(f: kotlin.Float): Int = f.hashCode()

    fun isNaN(f: kotlin.Float): kotlin.Boolean = f.isNaN()

    fun isInfinite(f: kotlin.Float): kotlin.Boolean = f.isInfinite()

    fun isFinite(f: kotlin.Float): kotlin.Boolean = f.isFinite()

    fun floatToIntBits(f: kotlin.Float): Int = f.toBits()

    fun floatToRawIntBits(f: kotlin.Float): Int = f.toRawBits()

    fun intBitsToFloat(bits: Int): kotlin.Float = kotlin.Float.fromBits(bits)

    fun min(f1: kotlin.Float, f2: kotlin.Float): kotlin.Float = kotlin.math.min(f1, f2)

    fun max(f1: kotlin.Float, f2: kotlin.Float): kotlin.Float = kotlin.math.max(f1, f2)

    fun sum(f1: kotlin.Float, f2: kotlin.Float): kotlin.Float = f1 + f2
  }
}
